pkg/controller/secret/secret_controller_test.go

```diff
@@ -0,0 +1,251 @@
+	//  group_by:
+	//  - job
+	//  routes:
+	//  - receiver: pagerduty
+	//    group_by:
+	//    - alertname
+	//    - severity
+	//	match_re:
+	//      namespace: ^openshift$|openshift-.*|default$|kube$|kube-.*|logging$
+	//   continue: true
+	//  - receiver: "null"
+	//    match:
+	//      alertname: Watchdog
+	//  group_wait: 30s
+	//  group_interval: 5m
+	//  repeat_interval: 12h
+	//receivers:
+	//- name: pagerduty
+	//  pagerduty_configs:
+	//  - send_resolved: true
+	//	routing_key: asdf1234567890
+	//	description: '{{ .CommonLabels.alertname }} {{ .CommonLabels.severity | toUpper
+	//    }} ({{ len .Alerts }})'
+	//    details:
+	//      component: '{{ .CommonLabels.alertname }}'
+	//      group: '{{ .CommonLabels.alertname }}'
+	//      link: '{{ .CommonAnnotations.link }}?'
+	//      num_firing: '{{ .Alerts.Firing | len }}'
+	//      num_resolved: '{{ .Alerts.Resolved | len }}'
+	//      resolved: '{{ template pagerduty.default.instances .Alerts.Resolved }}'
+	//- name: "null"
+	//templates: []
+	//`)
+
+	// Load the default AlertManager config into the fake cluster.
+	amconfigbyte := []byte(AlertManagerConfig)
+	reconciler := createReconciler()
+	createNamespace(reconciler, t)
+	createSecret("pd-secret", "PAGERDUTY_KEY", pdsecret, reconciler)
+	createAlertManagerConfig(amconfigbyte, reconciler)
+	req := createReconcileRequest(reconciler, "pd-secret")
+	want := getAlertManagerConfig(reconciler, req)
+
+	// Add the Pager Duty details to the fake alertmanager config.
+	pdconfig := &alertmanager.PagerdutyConfig{
+		NotifierConfig: alertmanager.NotifierConfig{VSendResolved: true},
+		RoutingKey:     pdsecret,
+		Description:    `{{ .CommonLabels.alertname }} {{ .CommonLabels.severity | toUpper }} ({{ len .Alerts }})`,
+		Details: map[string]string{
+			"link":         `{{ .CommonAnnotations.link }}?`,
+			"group":        `{{ .CommonLabels.alertname }}`,
+			"component":    `{{ .CommonLabels.alertname }}`,
+			"num_firing":   `{{ .Alerts.Firing | len }}`,
+			"num_resolved": `{{ .Alerts.Resolved | len }}`,
+			"resolved":     `{{ template pagerduty.default.instances .Alerts.Resolved }}`,
+		},
+	}
+	pdroute := &alertmanager.Route{
+		Continue: true,
+		Receiver: "pagerduty",
+		GroupByStr: []string{
+			"alertname",
+			"severity",
+		},
+		MatchRE: map[string]string{
+			"namespace": "^openshift$|openshift-.*|default$|kube$|kube-.*|logging$",
+		},
+	}
+	pdreceiver := &alertmanager.Receiver{
+		Name:             "pagerduty",
+		PagerdutyConfigs: []*alertmanager.PagerdutyConfig{pdconfig},
+	}
+	want.Receivers = append(want.Receivers, pdreceiver)
+	want.Route.Routes = append(want.Route.Routes, pdroute)
+	want.Global.PagerdutyURL = "https://events.pagerduty.com/v2/enqueue"
+
+	// Try to get the same result as `want`,
+	// using addPDSecretToAlertManagerConfig() and updateAlertManagerConfig().
+	defaultconfig := getAlertManagerConfig(reconciler, req)
+	addPDSecretToAlertManagerConfig(reconciler, req, &defaultconfig, pdsecret)
```

Comment:
Can you mock out the event to verify the controller?  Maybe it should be a separate test to check that you'll reconcile when a secret lands in the operator's namespace.  For example see https://github.com/openshift/dedicated-admin-operator/blob/master/pkg/controller/rolebinding/rolebinding_controller_test.go#L62-L74

Reply:
That would run the entire Reconcile function, which is where I had a lot of problems. This is actually the example I was working from initially where I got stuck.

That said, I've learned a lot about simulating the kubernetes objects I need since then. I can give it another go. This is essentially a functional test of the entire controller, which in my mind would be better to run on a real cluster. But I'm ok with running a fake version of it too, assuming I can get it working this time.

Reply:
Added! I used a table-driven test for this one, to show the reconcile event happening for a number of different secrets. Each one tests a logic path in the reconciler controller, which is somewhat useful, since it helped me catch an error that would have made the controller crash if alertmanager-main goes missing.